lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,565 @@
+            optimizer_init (Dict[str, Any]): The desired optimizer configuration.
+            lr_scheduler_init (Dict[str, Any]): The desired learning rate scheduler config
+            pl_lrs_cfg (Optional[Dict[str, Any]]): Defines custom overrides of pytorch lightning lr_scheduler defaults
+                defined in ``_get_default_scheduler_config``
+            model_cfg (Optional[Dict[str, Any]], optional): Defines overrides of the default model config. Defaults to
+                ``None``.
+            task_name (str, optional): The SuperGLUE task to execute, one of ``'rte'``, ``'boolq'``. Defaults to "rte".
+            experiment_tag (str, optional): The tag to use for the experiment and tensorboard logs. Defaults to
+                "default".
+        """
+        super().__init__()
+        self.optimizer_init = optimizer_init
+        self.lr_scheduler_init = lr_scheduler_init
+        self.pl_lrs_cfg = pl_lrs_cfg or {}
+        if task_name in TASK_NUM_LABELS.keys():
+            self.task_name = task_name
+        else:
+            self.task_name = DEFAULT_TASK
+            rank_zero_warn(f"Invalid task_name '{task_name}'. Proceeding with the default task: '{DEFAULT_TASK}'")
+        self.num_labels = TASK_NUM_LABELS[self.task_name]
+        self.save_hyperparameters()
+        self.experiment_id = f"{datetime.now().strftime('%Y%m%d_%H%M%S')}_{experiment_tag}"
+        self.model_cfg = model_cfg or {}
+        conf = AutoConfig.from_pretrained(model_name_or_path, num_labels=self.num_labels, local_files_only=False)
+        self.model = AutoModelForSequenceClassification.from_pretrained(model_name_or_path, config=conf)
+        self.model.config.update(self.model_cfg)  # apply model config overrides
+        self.metric = datasets.load_metric("super_glue", self.task_name, experiment_id=self.experiment_id)
+        self.no_decay = ["bias", "LayerNorm.weight"]
+        self.finetuningscheduler_callback = None
+
+    def forward(self, **inputs):
+        return self.model(**inputs)
+
+    def training_step(self, batch, batch_idx):
+        outputs = self(**batch)
+        loss = outputs[0]
+        return loss
+
+    def training_epoch_end(self, outputs: List[Any]) -> None:
+        loss = torch.stack([x["loss"] for x in outputs]).mean()
+        self.log("train_loss", loss, prog_bar=True, sync_dist=True)
+        if self.finetuningscheduler_callback:
+            self.log("finetuning_schedule_depth", float(self.finetuningscheduler_callback.curr_depth))
+
+    def validation_step(self, batch, batch_idx, dataloader_idx=0):
+        outputs = self(**batch)
+        val_loss, logits = outputs[:2]
+
+        if self.num_labels >= 1:
+            preds = torch.argmax(logits, axis=1)
+        elif self.num_labels == 1:
+            preds = logits.squeeze()
+
+        labels = batch["labels"]
+        return {"loss": val_loss, "preds": preds, "labels": labels}
+
+    def validation_epoch_end(self, outputs):
```

Comment:
I strongly recommend to log within the validation_step and let Lightning handles the step, epoch logging.

Reply:
Good point. Moved val_loss logging to validation_step.

Reply:
I would actually move all logging directly to the validation_step, so you don't even need to return anything. Once a large dataset, this could raise OOM.